spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -0,0 +1,584 @@
+    // in case it was modified
+    proposedUpgradeChannel.version = version
+
+    provableStore.set(timeoutPath(portIdentifier, channelIdentifier), upgradeTimeout)
+    provableStore.set(channelPath(portIdentifier, channelIdentifier), proposedUpgradeChannel)
+    privateStore.set(restorePath(portIdentifier, channelIdentifier), currentChannel)
+}
+```
+
+NOTE: It is up to individual implementations how they will provide access-control to the `ChanUpgradeInit` function. E.g. chain governance, permissioned actor, DAO, etc.
+Access control on counterparty should inform choice of timeout values, i.e. timeout value should be large if counterparty's `UpgradeTry` is gated by chain governance.
+
+```typescript
+function chanUpgradeTry(
+    portIdentifier: Identifier,
+    channelIdentifier: Identifier,
+    proposedUpgradeChannel: ChannelEnd,
+    timeoutHeight: Height,
+    timeoutTimestamp: uint64,
+    UpgradeTimeout: UpgradeTimeout,
+    proofChannel: CommitmentProof,
+    proofUpgradeTimeout: CommitmentProof,
+    proofHeight: Height
+) {
+    // current channel must be OPEN or UPGRADE_INIT (crossing hellos)
+    currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
+    abortTransactionUnless(currentChannel.state == OPEN || currentChannel.state == UPGRADE_INIT)
+
+    // abort transaction if an unmodifiable field is modified
+    // upgraded channel state must be in `UPGRADE_TRY`
+    // NOTE: Any added fields are by default modifiable.
+    abortTransactionUnless(
+        proposedUpgradeChannel.state == UPGRADE_TRY &&
+        proposedUpgradeChannel.counterpartyPortIdentifier == currentChannel.counterpartyPortIdentifier &&
+        proposedUpgradeChannel.counterpartyChannelIdentifier == currentChannel.counterpartyChannelIdentifier
+    )
+
+    // current ordering must be a valid ordering of packets
+    // in the proposed ordering
+    // e.g. ORDERED -> UNORDERED, ORDERED -> DAG
+    abortTransactionUnless(
+        currentChannel.ordering.subsetOf(proposedUpgradeChannel.ordering)
+    )
+
+    // verify proofs of counterparty state
+    abortTransactionUnless(verifyChannelState(currentChannel, proofHeight, proofChannel, currentChannel.counterpartyChannelIdentifier, counterpartyChannel))
+    abortTransactionUnless(verifyUpgradeTimeout(currentChannel, proofHeight, proofUpgradeTimeout, currentChannel.counterpartyChannelIdentifier, upgradeTimeout))
+
+    if currentChannel.state == UPGRADE_INIT {
+        // if there is a crossing hello, ie an UpgradeInit has been called on both channelEnds,
+        // then we must ensure that the proposedUpgrade by the counterparty is the same as the currentChannel
+        // except for the channel state (upgrade channel will be in UPGRADE_TRY and current channel will be in UPGRADE_INIT)
+        // if the proposed upgrades on either side are incompatible, then we will restore the channel and cancel the upgrade.
```

Comment:
Agreed this is a complicated case, and we could design it to handle this such that chainB goes to OPEN. But I think that is quite difficult to engineer and evaluate for correctness. It is better for the first implementation to just revert the channel in this case. This is an inconvenience when we reach this edge case, since we have to restart the handshake but I believe that is fine.

So we will start with:

chainA, UPGRADE_INIT, timeout 5
chainB, UPGRADE_INIT, timeout 10

at time 7, both chains attempt to go to TRY. chainA succeeds and chainB fails. So we have:

chainA, UPGRADE_TRY
chainB, OPEN (reverted and error receipt written. See pseudocode below where we restore channel if timeout has passed)

Then at a time < 10 relayer can send a `CancelUpgradeMsg` to chainA:

chainA, OPEN (reverted)
chainB, OPEN (reverted and error receipt written)